matrix_content_scanner/scanner/scanner.py

```diff
@@ -0,0 +1,190 @@
+            media_content = self._decrypt_file(media_content, metadata)
+
+        # Write the file to disk.
+        file_path = self._write_file_to_disk(media_path, media_content)
+
+        # Scan the file and see if the result is positive or negative.
+        exit_code = self._run_scan(file_path)
+        result = exit_code == 0
+
+        # Delete the file now that we've scanned it.
+        logger.info("Scan has finished, removing file")
+        removal_command_parts = self._removal_command.split()
+        removal_command_parts.append(file_path)
+        subprocess.run(removal_command_parts)
+
+        # Raise an error if the result isn't clean.
+        if result is False:
+            raise FileDirtyError()
+
+        return media
+
+    def _decrypt_file(self, body: bytes, metadata: JsonDict) -> bytes:
+        """Extract decryption information from the file's metadata and decrypt it.
+
+        Args:
+            body: The encrypted body of the file.
+            metadata: The part of the request that includes decryption information.
+
+        Returns:
+            The decrypted content of the file.
+
+        Raises:
+            ContentScannerRestError(400) if the decryption failed.
+        """
+        logger.info("File is encrypted, decrypting")
+
+        # At this point the schema should have been validated so we can pull these values
+        # out safely.
+        key = metadata["file"]["key"]["k"]
+        hash = metadata["file"]["hashes"]["sha256"]
+        iv = metadata["file"]["iv"]
+
+        # Decrypt the file.
+        try:
+            return decrypt_attachment(body, key, hash, iv)
+        except DecryptionError as e:
+            raise ContentScannerRestError(
+                http_status=400,
+                reason=ErrCode.FAILED_TO_DECRYPT,
+                info=e.message,
+            )
+
+    def _write_file_to_disk(self, media_path: str, body: bytes) -> str:
+        """Writes the given content to disk. The final file name will be a concatenation
+        of `temp_directory` and the media's `server_name/media_id` path.
+
+        Args:
+            media_path: The `server_name/media_id` path of the media we're processing.
+            body: The bytes to write to disk.
+
+        Returns:
+            The full path to the newly written file.
+
+        Raises:
+            FileDirtyError if the media path is malformed in a way that would cause the
+                file to be written outside the configured directory.
+        """
+        # Figure out the full absolute path for this file. Given _store_directory is
+        # already an absolute path using os.path.join is likely good enough, but we want
+        # to make sure there isn't any '..' etc in the full path, to make sure we don't
+        # try to write outside the directory.
+        full_path = os.path.abspath(os.path.join(self._store_directory, media_path))
+        if not full_path.startswith(self._store_directory):
```

Comment:
This seems slightly incorrect (probably not harmfully so, but still):

`"/blarrrr".startswith("/bla")` is true.

Are you aware of `pathlib` in modern Python versions? I find that much nicer for working with paths.

In this case you can probably write ...

```suggestion
        full_path = Path(self._store_directory).joinpath(media_path).resolve()
        if not full_path.is_relative_to(self._store_directory):
```

(I just found out you can also write `/` instead of `joinpath()` but ehh)

Reply:
> "/blarrrr".startswith("/bla") is true.

Gah, good catch!

> Are you aware of pathlib in modern Python versions? I find that much nicer for working with paths.

TIL, thanks! That looks indeed a bit nicer than `os.path`. Plus `resolve(strict=True)` provides a nice and easy way to figure out whether the directory provided as the store directory actually exists.

Reply:
Uuuuuugh `is_relative_to` was introduced in 3.9.